### PR TITLE
fix(no-eval): catch missing eval cases

### LINKED ISSUE #466: no-eval false negatives
es-lint flags the following three examples with the no-eval rule which deno lint does not:
```javascript
(0, eval)("var a = 0");

var foo = eval;
foo("var a = 0");

this.eval("var a = 0");
```

### PR DESCRIPTION
**Pull request by ross-weir**
Fixes #466 

Detect `eval` usage in the following situations:

```
this.eval();
var foo = eval;
(0, eval)();
((eval))();
```

### CODE REVIEW

File: src/rules/no_eval.rs [hunk 1]
```diff
@@ -59,23 +62,52 @@ impl<'c> NoEvalVisitor<'c> {
...
+          let ident_name = &ident.sym.as_ref();
+          self.maybe_add_diagnostic(ident_name, call_expr.span)
+        }
+        Expr::Member(member_expr) => {
+          let member_name = &member_expr.get_key().unwrap();
+          self.maybe_add_diagnostic(member_name, call_expr.span)
+        }
+        Expr::Paren(paren) => {
+          let paren_snippet =
+            &self.context.source_map.span_to_snippet(paren.span).unwrap();
```
Comment (ross-weir) [inline src/rules/no_eval.rs:107]: Couldn't figure out how to extract identifiers in the parens, let me know if there's a better way to do this!

Comment (magurotuna) [reply src/rules/no_eval.rs:107]: `paren` in `Expr::Paren(paren)` is `swc_ecma_ast::ParenExpr` type, which has `expr` field in it. This `expr` field is `Box<Expr>` type, and if it is equal to `eval`, it means that this `expr` looks like `Box<Expr::Ident>` where the field `sym`of `Expr::Ident` is `eval`.
Also you should probably take care of nested parens i.e. `((eval))();`

Comment (magurotuna) [reply src/rules/no_eval.rs:107]: For debugging, you may find `dbg!` so helpful. For example, if we add the following code in `Visit` implementation:

```rust
  fn visit_program(&mut self, p: &swc_ecmascript::ast::Program, _: &dyn Node) {
    dbg!(p);
    p.visit_children_with(self);
  }
```

then we can see what AST looks like.

```
[src/rules/no_eval.rs:69] p = Script(
    Script {
        span: Span {
            lo: BytePos(
                0,
            ),
            hi: BytePos(
                29,
            ),
            ctxt: #0,
        },
        body: [
            Expr(
                ExprStmt {
                    span: Span {
                        lo: BytePos(
                            0,
                        ),
                        hi: BytePos(
                            29,
                        ),
                        ctxt: #0,
                    },
                    expr: Call(
                        CallExpr {
                            span: Span {
                                lo: BytePos(
                                    0,
                                ),
                                hi: BytePos(
                                    29,
                                ),
                                ctxt: #0,
                            },
                            callee: Expr(
                                Paren(
                                    ParenExpr {
                                        span: Span {
                                            lo: BytePos(
                                                0,
                                            ),
                                            hi: BytePos(
                                                9,
                                            ),
                                            ctxt: #0,
                                        },
                                        expr: Seq(
                                            SeqExpr {
                                                span: Span {
                                                    lo: BytePos(
                                                        1,
                                                    ),
                                                    hi: BytePos(
                                                        8,
                                                    ),
                                                    ctxt: #0,
                                                },
                                                exprs: [
                                                    Lit(
                                                        Num(
                                                            Number {
                                                                span: Span {
                                                                    lo: BytePos(
                                                                        1,
                                                                    ),
                                                                    hi: BytePos(
                                                                        2,
                                                                    ),
                                                                    ctxt: #0,
                                                                },
                                                                value: 0.0,
                                                            },
                                                        ),
                                                    ),
                                                    Ident(
                                                        Ident {
                                                            span: Span {
                                                                lo: BytePos(
                                                                    4,
                                                                ),
                                                                hi: BytePos(
                                                                    8,
                                                                ),
                                                                ctxt: #1,
                                                            },
                                                            sym: Atom('eval' type=static),
                                                            type_ann: None,
                                                            optional: false,
                                                        },
                                                    ),
                                                ],
                                            },
                                        ),
                                    },
                                ),
                            ),
                            args: [
                                ExprOrSpread {
                                    spread: None,
                                    expr: Lit(
                                        Str(
                                            Str {
                                                span: Span {
                                                    lo: BytePos(
                                                        10,
                                                    ),
                                                    hi: BytePos(
                                                        28,
                                                    ),
                                                    ctxt: #0,
                                                },
                                                value: Atom('console.log(42);' type=dynamic),
                                                has_escape: false,
                                            },
                                        ),
                                    ),
                                },
                            ],
                            type_args: None,
                        },
                    ),
                },
            ),
        ],
        shebang: None,
    },
)
```

Comment (ross-weir) [reply src/rules/no_eval.rs]: That's extremely useful. Thanks a lot for the help!

File: src/rules/no_eval.rs [hunk 2]
```diff
@@ -1,10 +1,14 @@
...
 use super::Context;
 use super::LintRule;
+use crate::swc_util::StringRepr;
+use swc_common::Span;
 use swc_ecmascript::ast::CallExpr;
 use swc_ecmascript::ast::Expr;
 use swc_ecmascript::ast::ExprOrSuper;
+use swc_ecmascript::ast::VarDeclarator;
 use swc_ecmascript::visit::noop_visit_type;
+use swc_ecmascript::visit::swc_ecma_ast::ParenExpr;
```
Comment (magurotuna) [inline src/rules/no_eval.rs:11]:
```suggestion
use swc_ecmascript::ast::ParenExpr;
```

I'm not sure why `ParenExpr` is also exported from `swc_ecmascript::visit::swc_ecma_ast`, but if you don't have any special reason, it's better to import like this.

File: src/rules/no_eval.rs [hunk 3]
```diff
@@ -59,23 +63,61 @@ impl<'c> NoEvalVisitor<'c> {
   fn new(context: &'c mut Context) -> Self {
     Self { context }
   }
+
+  fn maybe_add_diagnostic(&mut self, source: &dyn StringRepr, span: Span) {
+    if source.string_repr().unwrap() == "eval" {
+      self.add_diagnostic(span);
+    }
```
Comment (magurotuna) [inline src/rules/no_eval.rs:70]: This `unwrap` seems a kind of dangerous to me as it will likely cause panics in real world. How about writing like this?

```suggestion
    if source.string_repr().as_deref() == Some("eval") {
      self.add_diagnostic(span);
    }
```

File: src/rules/no_eval.rs [hunk 4]
```diff
@@ -59,23 +63,61 @@ impl<'c> NoEvalVisitor<'c> {
...
-            call_expr.span,
-            "no-eval",
-            "`eval` call is not allowed",
-            "Remove the use of `eval`",
-          );
+      match expr.as_ref() {
+        Expr::Ident(ident) => self.maybe_add_diagnostic(ident, call_expr.span),
+        Expr::Member(member_expr) => {
+          self.maybe_add_diagnostic(member_expr, call_expr.span)
         }
```
Comment (magurotuna) [inline src/rules/no_eval.rs:118]: This could result in false positive. For example:

```ts
const obj = {
  eval: (val) => { console.log(val); },
};

obj.eval("yeah");
```

In this case calling `obj.eval` is totally fine, but your code will report it mistakenly.

**Review by magurotuna — CHANGES_REQUESTED** [review]

**Comment by magurotuna** [comment]
I'm referring to ESLint's implementation (https://github.com/eslint/eslint/blob/master/lib/rules/no-eval.js). Now I feel like `no-eval` rule is quite difficult to implement if we try to port it perfectly. For instance, in order to check if `this.eval(..)` is valid or not, we also need to check if `this` is the global object or not.

**Comment by ross-weir** [comment]
> I'm referring to ESLint's implementation (https://github.com/eslint/eslint/blob/master/lib/rules/no-eval.js). Now I feel like `no-eval` rule is quite difficult to implement if we try to port it perfectly. For instance, in order to check if `this.eval(..)` is valid or not, we also need to check if `this` is the global object or not.
> 
> I think what we have to do first is port test cases from ESLint(https://github.com/eslint/eslint/blob/master/tests/lib/rules/no-eval.js), which can help us figure out how this rule should be implemented.

Yeah this is quite a bit more complicated than I anticipated. I'm certainly happy to keep chipping away at it but it's going to take me a bit of time compared to you experienced folks :-)

How would you like me to proceed here? I'm thinking something like:

- Remove the `Member` expr false positive
- Port across the tests as `TODO` tests

Then at least the rule is improved slightly and we have something to aim for with the TODO tests going forward?

**Comment by magurotuna** [comment]
> - Remove the Member expr false positive

Yeah, I think it would be nice to remove this false positive, because "false negative" is better than "false positive" in terms of linter.

> - Port across the tests as TODO tests

This seems also good. Personally I like to write test cases first and let them pass one by one as I implement.

To be honest, I don't think we should port this rule perfectly from ESLint, because it is not so important rule and there are more edge cases compared to its importance - we are allowed to make a compromise more or less.
How much compromise we make could be shown by test cases.

**Comment by ross-weir** [comment]
> To be honest, I don't think we should port this rule perfectly from ESLint, because it is not so important rule and there are more edge cases compared to its importance - we are allowed to make a compromise more or less.
How much compromise we make could be shown by test cases.

Makes sense. I ported some tests that I _think_ shouldn't be so bad (I don't really know though, happy to remove/add cases), also added a test case for the false positive.

File: src/rules/no_eval.rs [hunk 5]
```diff
@@ -84,14 +124,42 @@ impl<'c> Visit for NoEvalVisitor<'c> {
...
+  #[test]
+  fn no_eval_invalid() {
+    assert_lint_err! {
+      NoEval,
+      "eval('123');": [{col: 0, message: MESSAGE, hint: HINT}],
+      "(0, eval)('var a = 0');": [{col: 4, message: MESSAGE, hint: HINT}],
+      "((eval))('var a = 0');": [{col: 2, message: MESSAGE, hint: HINT}],
+      "var foo = eval;": [{col: 4, message: MESSAGE, hint: HINT}],
+
+      // TODO
```
Comment (magurotuna) [inline src/rules/no_eval.rs:145]: Why not add a link to this PR `https://github.com/denoland/deno_lint/pull/490` here for other developers to understand the context?

**Comment by magurotuna** [comment]
@ross-weir 

> I ported some tests that I think shouldn't be so bad (I don't really know though, happy to remove/add cases), also added a test case for the false positive.

Great 👍  I think it's okay for the time being.

> Completed no-eval for #431 as part of this PR as I was adding new tests.

thanks!

File: src/rules/no_eval.rs (hunk 5 above)
Comment (kitsonk) [inline src/rules/no_eval.rs:145]:
```suggestion
      // TODO (see: https://github.com/denoland/deno_lint/pull/490)
```

**Review by magurotuna — APPROVED** [review]
LGTM

**Review by kitsonk — APPROVED** [review]
LGTM
